arrows/ffmpeg/ffmpeg_video_input.cxx

```diff
@@ -0,0 +1,834 @@
+::open( std::string video_name )
+{
+  // Close any currently opened file
+  this->close();
+
+  d->video_path = video_name;
+
+  {
+    std::lock_guard< std::mutex > lock(d->open_mutex);
+
+    if (!kwiversys::SystemTools::FileExists(d->video_path))
+    {
+      // Throw exception
+      throw kwiver::vital::file_not_found_exception(video_name, "File not found");
+    }
+
+    if (!d->open(video_name))
+    {
+      throw kwiver::vital::video_runtime_exception("Video stream open failed for unknown reasons");
+    }
+  }
+}
+
+
+// ------------------------------------------------------------------
+void
+ffmpeg_video_input
+::close()
+{
+  if (d->f_packet.data) {
+    av_free_packet(&d->f_packet);  // free last packet
+  }
+
+  if (d->f_frame)
+  {
+    av_freep(&d->f_frame);
+  }
+  d->f_frame = nullptr;
+
+  if (d->f_video_encoding && d->f_video_encoding->opaque)
+  {
+    av_freep(&d->f_video_encoding->opaque);
+  }
+
+  //d->num_frames_ = -2;
+  //is_->contig_memory_ = 0;
+  d->f_video_index = -1;
+  d->f_data_index = -1;
+  d->f_start_time = -1;
+  d->video_path = "";
+  //is_->metadata_.clear();
+  if (d->f_video_stream)
+  {
+    avcodec_close(d->f_video_stream ->codec);
+    d->f_video_stream = nullptr;
+  }
+  if (d->f_format_context)
+  {
+    avformat_close_input(&d->f_format_context);
+    d->f_format_context = nullptr;
+  }
+
+  d->f_video_encoding = nullptr;
+}
+
+// ------------------------------------------------------------------
+bool
+ffmpeg_video_input
+::next_frame( kwiver::vital::timestamp& ts,
+              uint32_t timeout )
+{
+  if (!d->is_opened())
+  {
+    throw vital::file_not_read_exception(d->video_path, "Video not open");
+  }
+
+  bool ret = d->advance();
+  if (ret)
+  {
+    ts.set_frame(d->frame_number());
+  }
+  return ret;
+}
+
+// ------------------------------------------------------------------
+bool ffmpeg_video_input::seek_frame(kwiver::vital::timestamp& ts,
+  kwiver::vital::timestamp::frame_t frame_number,
+  uint32_t timeout)
+{
+  LOG_INFO(this->logger(), "Seeking isn't supported yet");
+  return false;
+}
+
+
+// ------------------------------------------------------------------
+kwiver::vital::image_container_sptr
+ffmpeg_video_input
+::frame_image( )
+{
+  // Quick return if the stream isn't valid
+  if (!d->is_valid())
+  {
+    return nullptr;
+  }
+
+  AVCodecContext* enc = d->f_format_context->streams[d->f_video_index]->codec;
+
+  // If we have not already converted this frame, try to convert it
+
+  if (!d->current_image_memory && d->f_frame->data[0] != 0)
+  {
+    int width = enc->width;
+    int height = enc->height;
+
+    // If the pixel format is not recognized by then convert the data into RGB_24
+    ffmpeg_video_input::priv::pixel_format format = d->pixel_format_from_ffmpeg(enc->pix_fmt);
+    if (format == ffmpeg_video_input::priv::PIXEL_FORMAT_UNKNOWN)
+    {
+      int size = width * height * 3;
+      if (!d->current_image_memory || size != d->current_image_memory->size())
```

Comment:
We probably don't want to reuse the same image memory for decoding each frame.  This is great for efficiency, but will corrupt any image that an outside user is still holding on to.  At this point we should just allocate a new image for each new frame.  You can probably just make the image directly and let it allocate it's own memory and access the memory instead of needing to allocate the memory and building an image object around it.

If this memory allocation becomes a bottleneck, I recommend creating a special image memory allocator that builds a pool of image image memory objects such that the memory gets returned to the pool instead of deallocated when the last shared point to an image goes out of scope.  However, this optimization is complex and should be put off until it is needed.